docs/zenith/developer/go/457202-test-build-publish.md

```diff
@@ -0,0 +1,303 @@
+
+```shell
+dagger call test
+```
+
+Here's an example of the output you will see:
+
+```shell
+✔ dagger call test [58.42s]
+┃ PASS src/App.test.js
+┃   ✓ renders learn react link (101 ms)
+┃
+┃ Test Suites: 1 passed, 1 total
+┃ Tests:       1 passed, 1 total
+┃ Snapshots:   0 total
+┃ Time:        7.619 s
+┃ Ran all test suites.
+```
+
+:::tip
+This listing uses the `node` module's `Container.Run().Stderr()` method to explicitly specify the test command and print its error stream. As an alternative, the `node` module also exposes a `Test()` method, which executes the `npm run test` command and prints its output stream.
+:::
+
+## Step 4: Add a function to build the application
+
+If your application passes all its tests, the typical next step is to build it.
+
+Add a new `Build()` function that creates a production build of the example application:
+
+```go
+func (m *Mymod) Build(nodeVersion Optional[string]) *Directory {
+  return m.buildBase(nodeVersion).Build().Container().Directory("./build")
+}
+```
+
+This function does the following:
+
+- It calls the `buildBase()` function to obtain a `Container` with the application source code, Node.js runtime and cache volumes.
+- It calls the module's `Container.Build()` method, which returns a revised `Container` after setting the `npm run build` command to run in the container image.
+- It obtains a reference to the `build/` directory in the container with the `Container.Directory()` method. This method returns a `Directory` object.
+
+:::note
+The `npm run build` command is appropriate for a React application, but other applications are likely to use different commands. Modify your Dagger pipeline accordingly.
+:::
+
+Test the function by running it as below:
+
+```shell
+dagger call build
+```
+
+Here's an example of the output you will see:
+
+```shell
+✔ dagger call build [2m12.6s]
+┃ asset-manifest.json
+┃ favicon.ico
+┃ index.html
+┃ logo192.png
+┃ logo512.png
+┃ manifest.json
+┃ robots.txt
+┃ static
+```
+
+## Step 5: Add functions to publish the application image
+
+At this point, your Dagger module has functions to test and build the application. However, Dagger SDKs also have built-in support to publish container images to remote registries.
+
+One such registry is [ttl.sh](https://ttl.sh), an ephemeral Docker registry. The [Daggerverse](https://daggerverse.dev) already includes a `ttlsh` module to publish to this registry.
+
+1. Add the `ttlsh` module as a dependency in your module:
+
+  ```shell
+  dagger mod install github.com/shykes/daggerverse/ttlsh@16e40ec244966e55e36a13cb6e1ff8023e1e1473
+  ```
+
+1. Update the module and add new `Package()` and `Publish()` functions to copy the built application into an NGINX web server container and deliver the result to the registry:
+
+  ```go
+  func (m *Mymod) Package(nodeVersion Optional[string]) *Container {
+    return dag.Container().From("nginx:1.23-alpine").
+      WithDirectory("/usr/share/nginx/html", m.Build(nodeVersion))
```

Comment:
This one should have the `WithExposedPort` calls on it; the exposed ports are helpful for the service but they also result in those exposed ports being part of the published container image metadata.

Reply:
Modified